### PR TITLE
Worktree.Reset refactor and Soft, Merge, Hard and Mixed modes

### LINKED ISSUE #435: Worktree.Checkout:Force bug?
1.I clone a test Repository to /Users/snail/codes/gotest (current branch is master)
2.I delete a file "/Users/snail/codes/gotest/index.php"
3.use git command,git checkout master -f 
4. nothing error showed and index.php returned back,i hope so.
but if is use go-git do step 3 ,i got an error :remove /Users/snail/codes/gotest/index.php: no such file or directory
code snippet:
err = w.Checkout(&git.CheckoutOptions{
		Branch: plumbing.ReferenceName("refs/heads/master"),
		Force:  true,
	})
fmt.Println(err)
is it a bug? I confused .

### PR DESCRIPTION
**Pull request by mcuadros**
Fixes #435  #415

### CODE REVIEW

File: submodule.go [hunk 1]
```diff
@@ -90,7 +93,7 @@ func (s *Submodule) status(idx *index.Index) (*SubmoduleStatus, error) {
 		err = nil
 	}
 
-	return status, err
+	return status, nil
```
Comment (erizocosmico) [inline submodule.go]: Are we intentionally discarding the error here? If so, the previous if can be removed.

Comment (erizocosmico) [inline submodule.go]: Are we intentionally discarding the error here? If so, the previous if can be removed.

File: worktree.go [hunk 2]
```diff
@@ -270,65 +273,99 @@ func (w *Worktree) Reset(opts *ResetOptions) error {
...
+			name = ch.To.String()
+			e, err = t.FindEntry(name)
+			if err != nil {
+				return err
+			}
+		case merkletrie.Delete:
+			name = ch.From.String()
+		}
+
+		_, _ = idx.Remove(name)
```
Comment (erizocosmico) [inline worktree.go]: can this operation fail? If so, should we leave it unhandled?

File: worktree.go [hunk 3]
```diff
@@ -424,9 +485,7 @@ func (w *Worktree) checkoutChangeRegularFile(name string,
 ) error {
 	switch a {
 	case merkletrie.Modify:
-		if err := w.rmIndexFromFile(name, idx); err != nil {
-			return err
-		}
+		_, _ = idx.Remove(name)
```
Comment (erizocosmico) [inline worktree.go]: same, should we ignore this error?

File: worktree.go [hunk 4]
```diff
@@ -503,6 +556,7 @@ func (w *Worktree) checkoutFileSymlink(f *object.File) (err error) {
 }
 
 func (w *Worktree) addIndexFromTreeEntry(name string, f *object.TreeEntry, idx *index.Index) error {
+	_, _ = idx.Remove(name)
```
Comment (erizocosmico) [inline worktree.go]: same

File: worktree.go [hunk 5]
```diff
@@ -513,6 +567,7 @@ func (w *Worktree) addIndexFromTreeEntry(name string, f *object.TreeEntry, idx *
 }
 
 func (w *Worktree) addIndexFromFile(name string, h plumbing.Hash, idx *index.Index) error {
+	_, _ = idx.Remove(name)
```
Comment (erizocosmico) [inline worktree.go]: same

File: worktree.go (hunk 2 above)
Comment (mcuadros) [reply worktree.go]: the only error returned by Remove, is `ErrEntryNotFound` and we don't care about it.

**Pull request merged**